<a href="https://colab.research.google.com/github/hugo451/Prova-C-lculo-Num-rico/blob/main/Prova_C%C3%A1lculo_Num%C3%A9rico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Exercícios de Avaliação (Cálculo Numérico)

# André Hugo Ramalho Lopes - 20190019350

## Métodos

In [1]:
import sympy as sy
import pandas as pd
import numpy as np
import plotly.express as px
from sympy.abc import x

In [12]:
def bissec(a,b,f,error):
  tabela = pd.DataFrame(columns=['n','a','b','x','f(a)','f(b)','f(x)','|b-a|'])
  i=0
  while True:
    x= (a+b)/2 
    if f(a) * f(x) < 0:
        b = x
    else:
        a = x
    tabela.loc[i]=[i,a,b,x,f(a),f(b),f(x),abs(b-a)]
    i=i+1
    if ((abs(b-a)) < error) or (abs(f(x)) < error):
      break
  return x,i,tabela

In [3]:
def derivate(f,var):
  x=sy.Symbol(var)
  f_l = f.diff(x)
  fi = x - (f/f_l)
  fi = sy.lambdify(x,fi,"numpy") 
  return fi

In [4]:
def newton(x,f_s,f,var,error):
  fi=derivate(f_s,var)
  table = pd.DataFrame(columns=['n','x_n','f(x_n)','|x_n+1-x_n|'])
  i=0
  while True:
    fx=f(x)
    aux=fi(x)
    table.loc[i]=[i, x,f(x),abs(aux-x)]
    i+=1
    if abs(fx) < error  or abs(aux-x) < error:
        break
    x=aux
  return x,i,table

## Questão 1

Temos os pontos $x_0 = 0$, $y_0 = 5$, $x_f = 50$ e $y_f = 12$ e a equação $y=\frac{T_a}{w}cosh(\frac{w\cdot{x}}{T_a}) + y_0 -\frac{T_a}{w}$.

Substituindo as informações fornecidas podemos chegar na seguinte equação:

$$y=\frac{T_a}{w}cosh(\frac{w\cdot{x}}{T_a}) + y_0 -\frac{T_a}{w} <=>$$
$$\frac{T_a}{w}cosh(\frac{w\cdot{x}}{T_a}) + y_0 - y -\frac{T_a}{w} = 0 <=>$$
$$T_acosh(\frac{w\cdot{x}}{T_a}) + (y_0 - y)\cdot{w} -T_a = 0 <=>$$
$$T_acosh(\frac{10\cdot{x}}{T_a}) + (5 - y)\cdot{10} -T_a = 0 <=>$$
$$T_acosh(\frac{10\cdot{50}}{T_a}) + (5 - 12)\cdot{10} -T_a = 0 <=>$$
**$$T_acosh(\frac{500}{T_a}) -T_a -70= 0$$**

Para facilitar o uso das ferramentas chamaremos $T_a$ de x, então:
$$x\cdot{cosh(\frac{500}{x})} -x -70= 0$$

Consideraremos $f(x) = x\cdot{cosh(\frac{500}{x})} -x -70$, na qual devemos achar as raízes para encontrarmos a tensão horizontal necessária para equilibrar o cabo do exercício. Para isso, utilizaremos o método de newton para encontrarmos as raízes.



In [5]:
#Vamos definir a função f(x):

fx = x*sy.cosh(500/x) -x -70;fx

x*cosh(500/x) - x - 70

In [6]:
f=sy.lambdify(x,fx,'numpy')

In [7]:
x=np.linspace(1500,2000,1000)

In [8]:
fig = px.line (x = x,
               y = f(x),
               labels = {'y':'f(x)','x':'Valores de x'}, 
               template = 'presentation')
fig.show()

De acordo com o gráfico tiraremos valores de início para nosso método funcionar!

In [9]:
erro = 0.01
x,i,table=newton(1770,fx,f,'x',erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(table)

Solução: 1797.261176 
Iterações: 3


n          x_n    f(x_n)  |x_n+1-x_n|
0  0.0  1770.000000  1.092342    26.839777
1  1.0  1796.839777  0.016632     0.421399
2  2.0  1797.261176  0.000004     0.000101

Logo, a tensão horizontal necessária para equilibrar o peso do cabo é $T_a = 1797.261176$ 

## Questão 2

Na questão de número dois temos uma equação que simula a velocidade de ascenção de um foguete em um voo vertical:

$$v = u\cdot{ln(\frac{M_0}{M_0 - q\cdot{t}})} + g\cdot{t}$$

O problema pede para acharmos uma raíz contida no intervalo [6, 8] com tolerância de até 0,8% ($erro <= 0,008$).
Além disso, o problema nos fornece os seguintes dados:

*   $u = 200 m/s$
*   $M_0 = 1600 Kg$
*   $g = 9.8 m/s^2$
*   $q = 27 Kg/s$
*   $v = 100 m/s$

Logo, podemos achar a equação:

$$100 = 200\cdot{ln(\frac{1600}{1600 - 27\cdot{t}})} + 9.8\cdot{t}$$

$$200\cdot{ln(\frac{1600}{1600 - 27\cdot{t}})} + 9.8\cdot{t} - 100 = 0$$
-

Fazendo $t = x$, podemos dizer que $f(x) = 200\cdot{ln(\frac{1600}{1600 - 27\cdot{t}})} + 9.8\cdot{t} - 100$ e, dessa forma, acharemos a raíz dessa função no intervalo explicitado, então:

In [5]:
#Vamos definir a função f(x):

fx = 200*sy.log(1600/(1600 - 27*x)) + 9.8*x - 100;fx

9.8*x + 200*log(1600/(1600 - 27*x)) - 100

In [6]:
f=sy.lambdify(x,fx,'numpy')

In [7]:
#Fazendo os valores de x pertencerem ao intervalo pedido:
x=np.linspace(6,8,1000)

In [8]:
fig = px.line (x = x,
               y = f(x),
               labels = {'y':'f(x)','x':'Valores de x'}, 
               template = 'presentation')
fig.show()

De acordo com o gráfico tiraremos valores de início para nosso método funcionar!

In [13]:
a = 6
b = 8
erro = 0.008
x,i,tabela=bissec(a,b,f,erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 7.460938 
Iterações: 8


n         a         b         x      f(a)      f(b)      f(x)     |b-a|
0  0.0  7.000000  8.000000  7.000000 -6.259009  7.405154 -6.259009  1.000000
1  1.0  7.000000  7.500000  7.500000 -6.259009  0.563741  0.563741  0.500000
2  2.0  7.250000  7.500000  7.250000 -2.849945  0.563741 -2.849945  0.250000
3  3.0  7.375000  7.500000  7.375000 -1.143682  0.563741 -1.143682  0.125000
4  4.0  7.437500  7.500000  7.437500 -0.290116  0.563741 -0.290116  0.062500
5  5.0  7.437500  7.468750  7.468750 -0.290116  0.136776  0.136776  0.031250
6  6.0  7.453125  7.468750  7.453125 -0.076679  0.136776 -0.076679  0.015625
7  7.0  7.453125  7.460938  7.460938 -0.076679  0.030046  0.030046  0.007812

Podemos ver que a raíz da equação que foi requisitada é $t = 7.460938$

## Questão 3

Na questão de número três um carro passa por um buraco e, o armotecedor faz o carro entrar em um movimento amortecido ao longo do tempo t, precisamos determinar os tempos para as 3 primeiras vezes que o carro passa pelo x(t) = 0, seguindo a equação e os dados fornecidos, temos:

$$x(t) = x_0\cdot{e^{-n\cdot{t}}}\cdot{(cos(p\cdot{t}) + \frac{n}{p}\cdot{sen(p\cdot{t}))}}$$

-

$$n = \frac{c}{2m}$$

-

$$p = \sqrt{\frac{k}{m} - \frac{c^2}{4m^2}}$$

-

Então:

$$x(t) = x_0\cdot{e^{-\frac{c}{2m}\cdot{t}}}\cdot{(cos(\frac{\sqrt{4mk - c^2}}{2m}\cdot{t}) + \frac{c}{\sqrt{4mk - c^2}}\cdot{sen(\frac{\sqrt{4mk - c^2}}{2m}\cdot{t}))}}$$


Se,



*   $k = 1.25\cdot{10^9} g/s^2$
*   $c = 1.4\cdot{10^7} g/s$
*   $m = 1.2\cdot{10^6} g$

Logo:

$$x(t) = x_0\cdot{e^{-5.8333\cdot{t}}}\cdot{(cos(31.7433\cdot{t}) + 0.1837\cdot{sen(31.7433\cdot{t}))}}$$

Como queremos determinar raízes de uma função $x(t)$, podemos trocar as dependências e reformular a função para que ela fique mais tranquila de se interpretar e calcular com funções do python, daí faremos:

$$f(x) = x_0\cdot{e^{-5.8333\cdot{t}}}\cdot{(cos(31.7433\cdot{x}) + 0.1837\cdot{sen(31.7433\cdot{x}))}}$$

Usando o método da bissecção podemos achar as 3 primeiras raízes (t cuja nossa referência é zerada $x(t) = 0$) pedidas pelo problema. Supondo que o buraco leve nosso ponto de referência chegar ao nível $x_0 = -0.2m$, logo:

$$f(x) = -0.2\cdot{e^{-5.8333\cdot{t}}}\cdot{(cos(31.7433\cdot{x}) + 0.1837\cdot{sen(31.7433\cdot{x}))}}$$

Com isso...



In [5]:
#Vamos definir a função:

fx = -0.2*sy.exp(-5.8333*x)*(sy.cos(31.7433*x) + 0.1837*sy.sin(31.7433*x));fx

-0.2*(0.1837*sin(31.7433*x) + cos(31.7433*x))*exp(-5.8333*x)

In [6]:
f=sy.lambdify(x,fx,'numpy')

In [7]:
x=np.linspace(0.001,0.3,1000)

In [8]:
fig = px.line (x = x,
               y = f(x),
               labels = {'y':'f(x)','x':'Valores de x'}, 
               template = 'presentation')
fig.show()

De acordo com o gráfico tiraremos valores de início para nosso método funcionar!

In [9]:
a = 0.001
b = 0.1
erro = 0.001
x,i,tabela=bissec(a,b,f,erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 0.055141 
Iterações: 6


n         a         b         x      f(a)      f(b)      f(x)     |b-a|
0  0.0  0.050500  0.100000  0.050500 -0.022549  0.112219 -0.022549  0.049500
1  1.0  0.050500  0.075250  0.075250 -0.022549  0.077893  0.077893  0.024750
2  2.0  0.050500  0.062875  0.062875 -0.022549  0.033959  0.033959  0.012375
3  3.0  0.050500  0.056688  0.056688 -0.022549  0.006861  0.006861  0.006188
4  4.0  0.053594  0.056688  0.053594 -0.007617  0.006861 -0.007617  0.003094
5  5.0  0.055141  0.056688  0.055141 -0.000313  0.006861 -0.000313  0.001547

In [10]:
a = 0.1
b = 0.2
erro = 0.001
x,i,tabela=bissec(a,b,f,erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 0.153906 
Iterações: 7


n         a         b         x      f(a)      f(b)      f(x)     |b-a|
0  0.0  0.150000  0.200000  0.150000  0.011205 -0.062896  0.011205  0.050000
1  1.0  0.150000  0.175000  0.175000  0.011205 -0.044979 -0.044979  0.025000
2  2.0  0.150000  0.162500  0.162500  0.011205 -0.020581 -0.020581  0.012500
3  3.0  0.150000  0.156250  0.156250  0.011205 -0.005376 -0.005376  0.006250
4  4.0  0.153125  0.156250  0.153125  0.002777 -0.005376  0.002777  0.003125
5  5.0  0.153125  0.154688  0.154688  0.002777 -0.001338 -0.001338  0.001563
6  6.0  0.153906  0.154688  0.153906  0.000710 -0.001338  0.000710  0.000781

In [11]:
a = 0.2
b = 0.3
erro = 0.001
x,i,tabela=bissec(a,b,f,erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 0.253125 
Iterações: 5


n         a        b         x      f(a)      f(b)      f(x)     |b-a|
0  0.0  0.250000  0.30000  0.250000 -0.004715  0.035214 -0.004715  0.050000
1  1.0  0.250000  0.27500  0.275000 -0.004715  0.026143  0.026143  0.025000
2  2.0  0.250000  0.26250  0.262500 -0.004715  0.012868  0.012868  0.012500
3  3.0  0.250000  0.25625  0.256250 -0.004715  0.004488  0.004488  0.006250
4  4.0  0.253125  0.25625  0.253125 -0.000029  0.004488 -0.000029  0.003125

Encontramos os 3 primeiros tempos em que o ponto de referência volta ao ponto x(t) = 0, são eles:



*   $t_1 = 0.055141$
*   $t_2 = 0.153906$
*   $t_3 = 0.253125$



## Questão 4

O nível de oxigênio de um rio é medido pela equação:

$$C = 10 - 15\cdot{(e^{-0.1x} - e^{-0.5x})}$$

Para saber qual a distância rio abaixo onde o nível de oxigênio é igual a 4, temos que igualar a equação a 4:

$$4 = 10 - 15\cdot{(e^{-0.1x} - e^{-0.5x})}$$

Com isso,

$$6 - 15\cdot{(e^{-0.1x} - e^{-0.5x})} = 0$$

Podemos fazer a equação igual a $f(x)$ e determinar sua raíz, que nos dará o ponto onde o nível de oxigênio é igual a 4, então:

$$f(x) = 6 - 15\cdot{(e^{-0.1x} - e^{-0.5x})}$$

Utilizando o método de newton, acharemos a distância x:

In [4]:
#Definindo a função f(x):

fx = 6 - 15*(sy.exp(-0.1*x) - sy.exp(-0.5*x));fx

6 + 15*exp(-0.5*x) - 15*exp(-0.1*x)

In [5]:
f=sy.lambdify(x,fx,'numpy')

In [8]:
x=np.linspace(0,10,1000)

In [9]:
fig = px.line (x = x,
               y = f(x),
               labels = {'y':'f(x)','x':'Valores de x'}, 
               template = 'presentation')
fig.show()

Podemos ver que existem duas distâncias que rio abaixo podem ter nível de oxigênio igual a 4, então vamos subdividir o problema e aplicar duas vezes o método de newton.

In [12]:
erro = 0.001
x,i,tabela=newton(0,fx,f,'x',erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 1.579960 
Iterações: 5


n       x_n    f(x_n)  |x_n+1-x_n|
0  0.0  0.000000  6.000000     1.000000
1  1.0  1.000000  1.525399     0.477923
2  2.0  1.477923  0.224991     0.098328
3  3.0  1.576251  0.007895     0.003708
4  4.0  1.579960  0.000011     0.000005

In [13]:
erro = 0.001
x,i,tabela=newton(7,fx,f,'x',erro)
print('Solução: {0:f} \nIterações: {1:d}'.format(x,i))
display(tabela)

Solução: 8.870936 
Iterações: 3


n       x_n    f(x_n)  |x_n+1-x_n|
0  0.0  7.000000 -0.995819     1.920956
1  1.0  8.920956  0.026411     0.050019
2  2.0  8.870936 -0.000022     0.000042

Logo, podemos ver que os dois pontos rio abaixo que têm nível de oxigênio igual a 4 são:



*   $x_1 = 1.579960$
*   $x_2 = 8.870936$

